modules/RTC/RTCUtils.js

```diff
@@ -904,169 +902,161 @@ class RTCUtils extends Listenable {
+                // for the id of MediaStream. Let's just say that a number
+                // contains no special characters.
+                const id = stream.id;
+
+                // XXX The return statement is affected by automatic
+                // semicolon insertion (ASI). No line terminator is allowed
+                // between the return keyword and the expression.
+                return (
+                    typeof id === 'number'
+                        ? id
+                        : SDPUtil.filterSpecialChars(id));
+            };
+            this.getTrackID = function(track) {
+                return track.id;
+            };
 
-                return;
+            if (!webkitMediaStream.prototype.getVideoTracks) {
+                webkitMediaStream.prototype.getVideoTracks = function() {
+                    return this.videoTracks;
+                };
+            }
+            if (!webkitMediaStream.prototype.getAudioTracks) {
+                webkitMediaStream.prototype.getAudioTracks = function() {
+                    return this.audioTracks;
+                };
             }
+        } else if (browser.isEdge()) {
+            this.RTCPeerConnectionType = ortcRTCPeerConnection;
+            this.getUserMedia
+                = wrapGetUserMedia(
+                    navigator.mediaDevices.getUserMedia.bind(
+                        navigator.mediaDevices),
+                    true);
+            this.enumerateDevices = rawEnumerateDevicesWithCallback;
+            this.attachMediaStream
+                = wrapAttachMediaStream((element, stream) => {
+                    defaultSetVideoSrc(element, stream);
+                });
+
+            // ORTC does not generate remote MediaStreams so those are
+            // manually created by the ORTC shim. This means that their
+            // id (internally generated) does not match the stream id
+            // signaled into the remote SDP. Therefore, the shim adds a
+            // custom jitsiRemoteId property with the original stream id.
+            this.getStreamID = function(stream) {
+                const id = stream.jitsiRemoteId || stream.id;
 
-            this._initPCConstraints(options);
+                return SDPUtil.filterSpecialChars(id);
+            };
 
-            rtcReady = true;
-            eventEmitter.emit(RTCEvents.RTC_READY, true);
-            screenObtainer.init(
-                options, this.getUserMediaWithConstraints.bind(this));
+            // Remote MediaStreamTracks generated by ORTC (within a
+            // RTCRtpReceiver) have an internally/random id which does not
+            // match the track id signaled in the remote SDP. The shim adds
+            // a custom jitsi-id property with the original track id.
+            this.getTrackID = function(track) {
+                return track.jitsiRemoteId || track.id;
+            };
+        } else {
+            logger.error('Browser does not appear to be WebRTC-capable');
```

Comment:
From the point of view of a library consumer/client, I get the idea of a `Promise`, `resolve`, `reject`, or really just about any return value which differentiates between success and failure. But what do I do with a method which doesn't return any value and only logs an error message? I'm afraid I don't feel comfortable with such a library design. If `Promise` is not desired (e.g. because it goes into an asynchronous execution flow), then at least return a boolean or an error to tell me that the initialization has failed, or throw an exception upon failure.

Reply:
FTR: we talked this though and this is going to throw.